---
title: "Feature analysis Figure 3"
author: 'Zuber'
date: "11012024"
output:
  html_document:
    toc: yes
---

```{r "setup", echo=TRUE}
require("knitr") #loading of the knitr package
knitr::opts_knit$set(root.dir = "/Users/hzuber/Documents/PAPERS/Pouclet_2023/R_data_final/Figure3")# Setting working directory
```


```{R}
#needed packages
require(plyr)
require(reshape2)
require(ggplot2)
library (grid)
require(dplyr)
library(xlsx)
library(viridis)
```
# Import  the different lists


## List of loci identified by the differents methods
```{R}
library(data.table)
data_listGMUCT <- fread("/Users/hzuber/Documents/PAPERS/Pouclet_2023/R_data_final/List_GMUCT.txt", header=T, sep="\t")
data_listTRIBE <- fread("/Users/hzuber/Documents/PAPERS/Pouclet_2023/R_data_final/List_TRIBE.txt", header=T, sep="\t")

```

### Extract individual lists

The GMUCT list of loci with down regulated fragments
```{R}
GMUCT_DOWN <- data_listGMUCT[data_listGMUCT$Melt_Cat=="GMUCT_DOWN"]

print(paste("The lists used for the analysis for GMUCT contains: ",
            nrow(GMUCT_DOWN),
            " loci with down regulated fragments", sep=""))
```


The TRIBE list of loci with ADAR edition (TRIBE DNE1D153N)
```{R}
TRIBE_D153N <- data_listTRIBE[data_listTRIBE$Melt_Cat=="TRIBE_D153N_EDITED",]

print(paste("The lists used for the analysis for TRIBE contains: ",
            nrow(TRIBE_D153N),
            " loci with edition (TRIBE DNE1D153N)", sep=""))

```

## Import control lists to be used

List of transcript detected in RNAseq data of WT plants :
  - RNAseq seedling, Schiaffini et al., Plant Physiology 2022
  - RNAseq flowers, this paper

```{R}
Control <- fread("/Users/hzuber/Documents/PAPERS/Pouclet_2023/R_data_final/List_control.txt")
Seedling_control <- Control[Control$Tissue=="14_day_old_seedling"]
Flowers_control <- Control[Control$Tissue=="Flowers"]


print(paste("The lists used for the analysis for seedling control contains: ",
            nrow(Seedling_control),
            " loci detected in RNAseq", sep=""))


print(paste("The lists used for the analysis for flowers control contains: ",
            nrow(Flowers_control),
            " loci detected in RNAseq", sep=""))

```


## Import the list of DNE1 targets identified in Nagarajan et al, Plant Cell, 2023

```{R}
Known_target <- fread("/Users/hzuber/Documents/PAPERS/Pouclet_2023/R_data_final/List_knownDNEtarget.txt")

print(paste("The lists used for the analysis for known DNE1 targets contains: ",
            nrow(Known_target),
            " loci identified in Nagarajan et al, Plant Cell, 2023", sep=""))
```


# Import feature data to be used for comparison

## Import data table with mRNA, CDS, UTR lengths and intron number

TAIR10 annotation, representative model
```{R}
Feature <- fread("/Users/hzuber/Documents/PAPERS/Pouclet_2023/R_data_final/TAIR10_mRNAfeature_forrepresentativetranscript2.txt")
```

## Import lists of genes with G4 and uORF


The list of loci containing rG4 is from Yang et al. genome Biology 2020
The list of loci containing uORF is from Hu et al.  IEEE transactions on nanoscience 2016

```{R}
G4 <- fread("/Users/hzuber/Documents/PAPERS/Pouclet_2023/R_data_final/List_G4_targets.txt", header=FALSE)
colnames(G4) <- c("Gene_ID", "Cat")
Nb_of_G4_genes <- length(unique(G4$Gene_ID))


uORF <- fread("/Users/hzuber/Documents/PAPERS/Pouclet_2023/R_data/Annotation/Other_features/List_uORF.txt", header=FALSE)
colnames(uORF) <- c("Transcript_ID", "Cat")
uORF$Gene_ID <- gsub(".*(AT.......).*$", "\\1",uORF$Transcript_ID)

Nb_of_uORF_genes <- length(unique(uORF$Gene_ID))


print(paste(Nb_of_G4_genes, "transcripts with detected G4", sep=" "))
print(paste(Nb_of_uORF_genes, "transcripts with detected uORF", sep=" "))
```

Indicating in the feature table if G4 or uORF were detected
```{R}
Feature$uORF <- "no_uORF"
uORF_list <- uORF$Gene_ID
Feature[(Feature$Gene_ID %in% uORF_list),]$uORF <- "uORF"


Feature$G4 <- "no_G4"
G4_list <- G4$Gene_ID
Feature[(Feature$Gene_ID %in% G4_list),]$G4 <- "G4"
```


# Make a table combining feature information vs different lists

 
Make four table for the four lists

```{R}
table_GMUCT_DOWN <- Feature[(Feature$Gene_ID %in% GMUCT_DOWN$AGI),]
table_GMUCT_DOWN$List <- "GMUCT DOWN"

table_TRIBE_D153N <- Feature[(Feature$Gene_ID %in% TRIBE_D153N$AGI),]
table_TRIBE_D153N$List <- "TRIBE D153N"

table_TRIBE_GMUCT <- Feature[(Feature$Gene_ID %in% TRIBE_D153N$AGI & Feature$Gene_ID %in% GMUCT_DOWN$AGI),]
table_TRIBE_GMUCT$List <- "TRIBED153N&GMUCT"

table_known <- Feature[Feature$Gene_ID %in% Known_target$AGI,]
table_known$List <- "GMUCT Nagarajan"

table_CONTROL_S <- Feature[Feature$Gene_ID %in% Seedling_control$AGI,]
table_CONTROL_S$List <- "seedling"

table_CONTROL_F <- Feature[Feature$Gene_ID %in% Flowers_control$AGI,]
table_CONTROL_F$List <- "flowers"

The_table <- rbind(table_GMUCT_DOWN, table_TRIBE_D153N, table_TRIBE_GMUCT, table_known, table_CONTROL_S, table_CONTROL_F)

Orderforgraph <- c("GMUCT DOWN", "TRIBE D153N","TRIBED153N&GMUCT", "GMUCT Nagarajan", "seedling", "flowers")
The_table$List <- factor(The_table$List, levels=Orderforgraph)
```



To check : Count number of genes per category
```{R}
freq_gene3 <-  The_table %>% 
                      group_by(List, Gene_ID) %>% 
                      summarise(n=n()) 


Count3 <- freq_gene3 %>% 
                      group_by(List) %>% 
                      summarise(n=n()) 


write.table(Count3, "Number_of_genes_by_category.txt", sep='\t', row.names=FALSE, quote=FALSE)
```

# Compute proportion of genes for uORF and rGA according to the list

```{R}
#uORF
The_table_freq_uORF <- The_table %>% 
  group_by(uORF, List) %>% 
  summarise(count = n()) %>% 
  group_by(List) %>%
  mutate(total=sum(count), Percent=100*count/sum(count))

The_table_freq_uORF$uORF <- factor(The_table_freq_uORF$uORF, levels=c("no_uORF", "uORF"))
The_table_freq_uORF$List <- factor(The_table_freq_uORF$List, levels=Orderforgraph)

The_table_freq_uORF_2 <- The_table_freq_uORF[The_table_freq_uORF$uORF=="uORF",]

#rG4
The_table_freq_G4 <- The_table %>% 
  group_by(G4, List) %>% 
  summarise(count = n()) %>% 
  group_by(List) %>%
  mutate(total=sum(count), Percent=100*count/sum(count))
The_table_freq_G4$G4 <- factor(The_table_freq_G4$G4, levels=c("no_G4", "G4"))
The_table_freq_G4$List <- factor(The_table_freq_G4$List, levels=Orderforgraph)

The_table_freq_G4_2 <- The_table_freq_G4[The_table_freq_G4$G4=="G4",]
``` 

# Make graphs
```{R}
my_theme <- theme(panel.spacing = unit(0.1, "lines"),
                  plot.title = element_text(hjust = 0.5, size=10, face = "bold"),
                  panel.background = element_blank(),
                  panel.grid.major.y = element_line(size = 0.2, linetype = 'dashed', colour = "gray70"),
                  panel.border = element_rect(colour="black",fill=NA,size=0.5),
                  legend.position = "none",
                  axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=6),
                  axis.title = element_blank(),
                  strip.text.x = element_blank()) 
```


```{R}
p_Intron <- ggplot(The_table, aes(x=List, y=nbre_intron, fill=List)) + 
            geom_boxplot(outlier.shape = NA, width=0.75) +
            scale_y_continuous(breaks=seq(0,20,by=2)) +
            coord_cartesian(ylim = c(0, 20)) +
            scale_fill_viridis(discrete = TRUE) +
            my_theme +
            ggtitle("Intron number")


p_CDS_length <- ggplot(The_table, aes(x=List, y=CDS, fill=List)) + 
            geom_boxplot(outlier.shape = NA, width=0.75) +
            scale_y_continuous(breaks=seq(0,5000,by=1000)) +
            coord_cartesian(ylim = c(0, 5000)) +
            scale_fill_viridis(discrete = TRUE) +
            my_theme +
                      ggtitle("CDS length")


p_a3UTR_length <- ggplot(The_table, aes(x=List, y=a3UTR_length, fill=List)) + 
            geom_boxplot(outlier.shape = NA, width=0.75) +
            scale_y_continuous(breaks=seq(0,600,by=100)) +
            coord_cartesian(ylim = c(0, 600)) +
            scale_fill_viridis(discrete = TRUE) +
            my_theme +
                      ggtitle("3' UTR length")

p_a5UTR_length <- ggplot(The_table, aes(x=List, y=a5UTR_length, fill=List)) + 
            geom_boxplot(outlier.shape = NA, width=0.75) +
            scale_y_continuous(breaks=seq(0,600,by=100)) +
            coord_cartesian(ylim = c(0, 600)) +
            scale_fill_viridis(discrete = TRUE) +
            my_theme +
                      ggtitle("5' UTR length")

p_uORF <- ggplot(The_table_freq_uORF_2, aes(x=List, y=Percent, fill=List)) + 
            geom_bar(stat="identity", width=0.75, color="black") +
            scale_fill_viridis(discrete = TRUE) +
            my_theme +
                      ggtitle("Translated uORFs Ribo-seq")

p_G4 <- ggplot(The_table_freq_G4_2, aes(x=List, y=Percent, fill=List)) + 
            geom_bar(stat="identity", width=0.75, color="black") +
            scale_fill_viridis(discrete = TRUE) +
            my_theme +
                      ggtitle("RNA-G4 in rG4-seq")


library(gridExtra)
p <- grid.arrange(p_CDS_length, p_a5UTR_length, p_uORF, p_Intron, p_a3UTR_length, p_G4, ncol=3, nrow = 2, widths=c(5, 5, 5), heights=c(4, 4))

ggsave(filename ="Feature_analysis_Figure3.pdf", p, width = 8, height = 6, dpi = 300)
```



# Statistical analysis

## Wilcoxon Rank Sum test test for panel C

Function for multicomparison
```{R}
tri.to.squ<-function(x)
{
rn<-row.names(x)
cn<-colnames(x)
an<-unique(c(cn,rn))
myval<-x[!is.na(x)]
mymat<-matrix(1,nrow=length(an),ncol=length(an),dimnames=list(an,an))
for(ext in 1:length(cn))
{
 for(int in 1:length(rn))
 {
 if(is.na(x[row.names(x)==rn[int],colnames(x)==cn[ext]])) next
 mymat[row.names(mymat)==rn[int],colnames(mymat)==cn[ext]]<-x[row.names(x)==rn[int],colnames(x)==cn[ext]]
 mymat[row.names(mymat)==cn[ext],colnames(mymat)==rn[int]]<-x[row.names(x)==rn[int],colnames(x)==cn[ext]]
 }
  
}
return(mymat)
}
library(multcompView)
```

Function for statistical analysis
```{R}
my_wilcoxon_test <- function(x, quantitative_variable, i) {
  
  #Wilcoxon test
  test <- pairwise.wilcox.test(quantitative_variable,x$List,p.adj="fdr", exact = FALSE, paired = FALSE) 
  #reformat output
  mymat_Sorenson<-tri.to.squ(test$p.value)
  #multicomparison
  mymat_Sorenson_test<-multcompLetters(mymat_Sorenson,compare="<=",threshold=0.001,Letters=letters)
  
  Output <- cbind(mymat_Sorenson, Letters=mymat_Sorenson_test$Letters) 
  print(paste(i, test$method, sep=" "))
  write.table(Output, paste("Statistics/Stats_0.001_", i, ".txt", sep=""), sep="\t", quote=FALSE, col.names = NA)
}
```


Tests for intron, 5utr, 3utr, mRNA length
```{R}
Stat_nbre_intron <- my_wilcoxon_test(The_table, The_table$nbre_intron, "nbre_intron")
Stat_CDSlength <- my_wilcoxon_test(The_table, The_table$CDS, "CDS_length")
Stat_a5UTR_length <- my_wilcoxon_test(The_table, The_table$a5UTR_length, "a5UTR_length")
Stat_a3UTR_length <- my_wilcoxon_test(The_table, The_table$a5UTR_length, "a3UTR_length")
```


##  Two-samples z-test of proportions for loci containing rG4 and uORF
Compare the proportion of survived between groups
 2-sample test for equality of proportions with continuity correction referred as a z-test
https://rpkgs.datanovia.com/rstatix/reference/prop_test.html

```{R}
library(rstatix)

uORF_conting <- xtabs(count ~ uORF + List, data=The_table_freq_uORF)
uORF_results <- pairwise_prop_test(uORF_conting)

G4_conting <- xtabs(count ~ G4 + List, data=The_table_freq_G4)
G4_results <- pairwise_prop_test(G4_conting)

#Export contingency table
write( "Contingency table \n ", file = "Statistics/Stats_uORF_results.txt")
write( "Contingency table \n ", file = "Statistics/Stats_G4_results.txt")

write.table(uORF_conting, "Statistics/Stats_uORF_results.txt", sep="\t", quote=FALSE, col.names = NA, append=T)
write.table(G4_conting, "Statistics/Stats_G4_results.txt", sep="\t", quote=FALSE, col.names = NA, append=T)
```


```{R}
write( "\n \n Statistic results \n ", file = "Statistics/Stats_uORF_results.txt", append = T)
write.table(uORF_results, "Statistics/Stats_uORF_results.txt", append = T, sep="\t", quote=FALSE, row.names = F)
uORF_results$Comparison <- paste(uORF_results$group1, uORF_results$group2, sep="-")
dif4 <- uORF_results$p.adj
names(dif4) <- uORF_results$Comparison
exp_letters1 <- multcompLetters(dif4, compare="<=",threshold=0.001)
df <- as.data.frame(exp_letters1$Letters)

df$List <-row.names(df)

write( "\n \n Statistic letters (p-value <= 0.001) \n ", file = "Statistics/Stats_uORF_results.txt", append = T)
write.table( df,
              file = "Statistics/Stats_uORF_results.txt", 
              append = T,
              row.names = F,
             col.names = F,
              na="",
             sep="\t",
              quote = F)

```

```{R}
write( "\n \n Statistic results \n ", file = "Statistics/Stats_G4_results.txt", append = T)
write.table(G4_results, "Statistics/Stats_G4_results.txt", append = T, sep="\t", quote=FALSE, row.names = F)
G4_results$Comparison <- paste(G4_results$group1, G4_results$group2, sep="-")
dif4 <- G4_results$p.adj
names(dif4) <- G4_results$Comparison
exp_letters1 <- multcompLetters(dif4, compare="<=",threshold=0.001)

df <- as.data.frame(exp_letters1$Letters)

df$List <-row.names(df)

write( "\n \n Statistic letters (p-value <= 0.001)\n ", file = "Statistics/Stats_G4_results.txt", append = T)
write.table( df,
              file = "Statistics/Stats_G4_results.txt", 
              append = T,
              row.names = F,
             col.names = F,
              na="",
             sep="\t",
              quote = F)

```


```{R}
sessionInfo()
```
